crash on compilable code: assertion failed: orphan poly parameter: PolyParam(T)

scalac can compile but dotc crashes.


$ cat abc.scala
```
case class B[T](b: List[Int]) {
  var s: B[Int] = _
  s copy ()
}
```

$ dotc abc.scala
```
error while pickling type T
error when pickling type T
error when pickling tree T
error when pickling tree $1$.copy[T]
error when pickling tree $1$.copy[T]($1$.copy$default$1[T])
error when pickling tree {
  val $1$: B[Int] = this.s
  $1$.copy[T]($1$.copy$default$1[T])
}
error when pickling tree [T](b: List[Int]) extends Object() with _root_.scala.Product1[List[Int]] {
  T
  private[this] type T = T
  val b: scala.collection.immutable.List[Int]
  var s: B[Int] = _
  <accessor> def s_=(x$1: B[Int]): Unit = ()
  {
    val $1$: B[Int] = this.s
    $1$.copy[T]($1$.copy$default$1[T])
  }
  def copy[T](b: scala.collection.immutable.List[Int]): B[T] = new B[T](b)
  def copy$default$1[T]: scala.collection.immutable.List[Int] @uncheckedVariance
     =
  B.this.b: scala.collection.immutable.List[Int] @uncheckedVariance
  def isDefined: Boolean = true
  def _1: scala.collection.immutable.List[Int] = this.b
  override def hashCode(): Int =
    {
      var acc: Int = -889275714
      acc =
        scala.runtime.Statics$#mix(acc, scala.runtime.Statics$#anyHash(B.this.b)
          )
      scala.runtime.Statics$#finalizeHash(acc, 1)
    }
  override def equals(x$0: Any): Boolean =
    this.eq(x$0.asInstanceOf[Object]).||(
      x$0 match
        {
          case x$0 @ _: B => this.b.==(x$0.b)
          case _: Any => false
        }
    )
  override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
  override def canEqual(that: Any): Boolean = that.isInstanceOf[B]
  override def productPrefix: String = "B"
}
error when pickling tree @scala.annotation.internal.SourceFile("abc.scala") case class B[T](b: List[Int])
   extends
 Object() with _root_.scala.Product1[List[Int]] {
  type T
  private[this] type T = T
  val b: scala.collection.immutable.List[Int]
  var s: B[Int] = _
  <accessor> def s_=(x$1: B[Int]): Unit = ()
  {
    val $1$: B[Int] = this.s
    $1$.copy[T]($1$.copy$default$1[T])
  }
  def copy[T](b: scala.collection.immutable.List[Int]): B[T] = new B[T](b)
  def copy$default$1[T]: scala.collection.immutable.List[Int] @uncheckedVariance
     =
  B.this.b: scala.collection.immutable.List[Int] @uncheckedVariance
  def isDefined: Boolean = true
  def _1: scala.collection.immutable.List[Int] = this.b
  override def hashCode(): Int =
    {
      var acc: Int = -889275714
      acc =
        scala.runtime.Statics$#mix(acc, scala.runtime.Statics$#anyHash(B.this.b)
          )
      scala.runtime.Statics$#finalizeHash(acc, 1)
    }
  override def equals(x$0: Any): Boolean =
    this.eq(x$0.asInstanceOf[Object]).||(
      x$0 match
        {
          case x$0 @ _: B => this.b.==(x$0.b)
          case _: Any => false
        }
    )
  override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
  override def canEqual(that: Any): Boolean = that.isInstanceOf[B]
  override def productPrefix: String = "B"
}
error when pickling tree package <empty> {
  @scala.annotation.internal.SourceFile("abc.scala") case class B[T](
    b: List[Int]
  ) extends Object() with _root_.scala.Product1[List[Int]] {
    type T
    private[this] type T = T
    val b: scala.collection.immutable.List[Int]
    var s: B[Int] = _
    <accessor> def s_=(x$1: B[Int]): Unit = ()
    {
      val $1$: B[Int] = this.s
      $1$.copy[T]($1$.copy$default$1[T])
    }
    def copy[T](b: scala.collection.immutable.List[Int]): B[T] = new B[T](b)
    def copy$default$1[T]:
      scala.collection.immutable.List[Int] @uncheckedVariance
     = B.this.b: scala.collection.immutable.List[Int] @uncheckedVariance
    def isDefined: Boolean = true
    def _1: scala.collection.immutable.List[Int] = this.b
    override def hashCode(): Int =
      {
        var acc: Int = -889275714
        acc =
          scala.runtime.Statics$#mix(acc,
            scala.runtime.Statics$#anyHash(B.this.b)
          )
        scala.runtime.Statics$#finalizeHash(acc, 1)
      }
    override def equals(x$0: Any): Boolean =
      this.eq(x$0.asInstanceOf[Object]).||(
        x$0 match
          {
            case x$0 @ _: B => this.b.==(x$0.b)
            case _: Any => false
          }
      )
    override def toString(): String = scala.runtime.ScalaRunTime._toString(this)
    override def canEqual(that: Any): Boolean = that.isInstanceOf[B]
    override def productPrefix: String = "B"
  }
  final lazy module val B: B$ = new B$()
  @scala.annotation.internal.SourceFile("abc.scala") final module class B$()
     extends
   Object() { this: B.type =>
    def apply[T](b: scala.collection.immutable.List[Int]): B[T] = new B[T](b)
    def unapply[T](x$1: B[T]): B[T] = x$1
  }
}

exception occurred while compiling abc.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: orphan poly parameter: PolyParam(T)
	at scala.Predef$.assert(Predef.scala:165)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:269)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:145)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:530)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:305)
	at dotty.tools.dotc.core.tasty.TreePickler$$anonfun$pickleTree$2$$anonfun$apply$mcV$sp$4.apply(TreePickler.scala:390)
	at dotty.tools.dotc.core.tasty.TreePickler$$anonfun$pickleTree$2$$anonfun$apply$mcV$sp$4.apply(TreePickler.scala:390)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at dotty.tools.dotc.core.tasty.TreePickler$$anonfun$pickleTree$2.apply$mcV$sp(TreePickler.scala:390)
	at dotty.tools.dotc.core.tasty.TreePickler.dotty$tools$dotc$core$tasty$TreePickler$$withLength(TreePickler.scala:30)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:388)
<snipped>
```

Thanks @helloqirun , report confirmed.

Compile with `-Ycheck:all` exposes the orphan param problem after frontend before pickling:

```
Exception in thread "main" java.lang.AssertionError: assertion failed: orphan param: PolyParam(T)
	at scala.Predef$.assert(Predef.scala:165)
	at dotty.tools.dotc.transform.TreeChecker$Checker$$anon$1.apply(TreeChecker.scala:296)
	at dotty.tools.dotc.transform.TreeChecker$Checker$$anon$1.apply(TreeChecker.scala:298)
	at dotty.tools.dotc.transform.TreeChecker$Checker.checkNoOrphans(TreeChecker.scala:304)
```

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -2862,14 +2862,14 @@ object Types {
    *
    *  @param  origin        The parameter that's tracked by the type variable.
    *  @param  creatorState  The typer state in which the variable was created.
-   *  @param  owningTree    The function part of the TypeApply tree tree that introduces
-   *                        the type variable.
+   *  @param  bindingTree   The TypeTree which introduces the type variable, or EmptyTree
+   *                        if the type variable does not correspond to a source term.
    *  @paran  owner         The current owner if the context where the variable was created.
    *
    *  `owningTree` and `owner` are used to determine whether a type-variable can be instantiated
    *  at some given point. See `Inferencing#interpolateUndetVars`.
    */
-  final class TypeVar(val origin: PolyParam, creatorState: TyperState, val owningTree: untpd.Tree, val owner: Symbol) extends CachedProxyType with ValueType {
+  final class TypeVar(val origin: PolyParam, creatorState: TyperState, val bindingTree: untpd.Tree, val owner: Symbol) extends CachedProxyType with ValueType {
 
     /** The permanent instance type of the variable, or NoType is none is given yet */
     private[core] var inst: Type = NoType

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -258,6 +258,28 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
         // method could be constructed as the definition site of the type variable for
         // that default constructor. This would interpolate type variables too early,
         // causing lots of tests (among them tasty_unpickleScala2) to fail.
+        //
+        // The test case is in i1757.scala. Here we have a variable `s` and a method `cpy`
+        // defined like this:
+        //
+        //      var s
+        //      def cpy[X](b: List[Int] = b): B[X] = new B[X](b)
+        //
+        // The call `s.cpy()` then gets expanded to
+        //
+        //      { val $1$: B[Int] = this.s
+        //        $1$.cpy[X']($1$.cpy$default$1[X']
+        //      }
+        //
+        // A type variable gets interpolated if it does not appear in the type
+        // of the current tree and the current tree contains the variable's "definition".
+        // Previously, the polymorphic function tree to which the variable was first added
+        // was taken as the variable's definition. But that fails here because that
+        // tree was `s.cpy` but got transformed into `$1$.cpy`. We now take the type argument
+        // [X'] of the variable as its definition tree, which is more robust. But then
+        // it's crucial that the type tree is not copied directly as argument to
+        // `cpy$default$1`. If it was, the variable `X'` would already be interpolated
+        // when typing the default argument, which is too early.
         spliceMeth(meth, fn).appliedToTypes(targs.tpes)
       case _ => meth
     }

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -250,8 +250,15 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
 
     /** Splice new method reference into existing application */
     def spliceMeth(meth: Tree, app: Tree): Tree = app match {
-      case Apply(fn, args) => Apply(spliceMeth(meth, fn), args)
-      case TypeApply(fn, targs) => TypeApply(spliceMeth(meth, fn), targs)
+      case Apply(fn, args) =>
+        spliceMeth(meth, fn).appliedToArgs(args)
+      case TypeApply(fn, targs) =>
+        // Note: It is important that the type arguments `targs` are passed in new trees
+        // instead of being spliced in literally. Otherwise, a type argument to a default
+        // method could be constructed as the definition site of the type variable for
+        // that default constructor. This would interpolate type variables too early,
+        // causing lots of tests (among them tasty_unpickleScala2) to fail.
+        spliceMeth(meth, fn).appliedToTypes(targs.tpes)
       case _ => meth
     }
 
@@ -333,7 +340,7 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
             val getter = findDefaultGetter(n + numArgs(normalizedFun))
             if (getter.isEmpty) missingArg(n)
             else {
-              addTyped(treeToArg(spliceMeth(getter withPos appPos, normalizedFun)), formal)
+              addTyped(treeToArg(spliceMeth(getter withPos normalizedFun.pos, normalizedFun)), formal)
               matchArgs(args1, formals1, n + 1)
             }
           }

--- a/compiler/src/dotty/tools/dotc/typer/Inferencing.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Inferencing.scala
@@ -216,10 +216,10 @@ object Inferencing {
   def interpolateUndetVars(tree: Tree, ownedBy: Symbol)(implicit ctx: Context): Unit = {
     val constraint = ctx.typerState.constraint
     val qualifies = (tvar: TypeVar) =>
-      (tree contains tvar.owningTree) || ownedBy.exists && tvar.owner == ownedBy
+      (tree contains tvar.bindingTree) || ownedBy.exists && tvar.owner == ownedBy
     def interpolate() = Stats.track("interpolateUndetVars") {
       val tp = tree.tpe.widen
-      constr.println(s"interpolate undet vars in ${tp.show}, pos = ${tree.pos}, mode = ${ctx.mode}, undets = ${constraint.uninstVars map (tvar => s"${tvar.show}@${tvar.owningTree.pos}")}")
+      constr.println(s"interpolate undet vars in ${tp.show}, pos = ${tree.pos}, mode = ${ctx.mode}, undets = ${constraint.uninstVars map (tvar => s"${tvar.show}@${tvar.bindingTree.pos}")}")
       constr.println(s"qualifying undet vars: ${constraint.uninstVars filter qualifies map (tvar => s"$tvar / ${tvar.show}")}, constraint: ${constraint.show}")
 
       val vs = variances(tp, qualifies)

--- a/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
@@ -353,20 +353,23 @@ object ProtoTypes {
    *  Also, if `owningTree` is non-empty, add a type variable for each parameter.
    *  @return  The added polytype, and the list of created type variables.
    */
-  def constrained(pt: PolyType, owningTree: untpd.Tree)(implicit ctx: Context): (PolyType, List[TypeVar]) = {
+  def constrained(pt: PolyType, owningTree: untpd.Tree)(implicit ctx: Context): (PolyType, List[TypeTree]) = {
     val state = ctx.typerState
     assert(!(ctx.typerState.isCommittable && owningTree.isEmpty),
       s"inconsistent: no typevars were added to committable constraint ${state.constraint}")
 
-    def newTypeVars(pt: PolyType): List[TypeVar] =
+    def newTypeVars(pt: PolyType): List[TypeTree] =
       for (n <- (0 until pt.paramNames.length).toList)
-      yield new TypeVar(PolyParam(pt, n), state, owningTree, ctx.owner)
+      yield {
+        val tt = new TypeTree().withPos(owningTree.pos)
+        tt.withType(new TypeVar(PolyParam(pt, n), state, tt, ctx.owner))
+      }
 
     val added =
       if (state.constraint contains pt) pt.newLikeThis(pt.paramNames, pt.paramBounds, pt.resultType)
       else pt
     val tvars = if (owningTree.isEmpty) Nil else newTypeVars(added)
-    ctx.typeComparer.addToConstraint(added, tvars)
+    ctx.typeComparer.addToConstraint(added, tvars.tpes.asInstanceOf[List[TypeVar]])
     (added, tvars)
   }
 

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -1962,12 +1962,12 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
           if (pt.isInstanceOf[PolyProto]) tree
           else {
             var typeArgs = tree match {
-              case Select(qual, nme.CONSTRUCTOR) => qual.tpe.widenDealias.argTypesLo
+              case Select(qual, nme.CONSTRUCTOR) => qual.tpe.widenDealias.argTypesLo.map(TypeTree)
               case _ => Nil
             }
             if (typeArgs.isEmpty) typeArgs = constrained(poly, tree)._2
             convertNewGenericArray(
-              adaptInterpolated(tree.appliedToTypes(typeArgs), pt, original))
+              adaptInterpolated(tree.appliedToTypeTrees(typeArgs), pt, original))
           }
         case wtp =>
           pt match {

--- a/tests/pos/i1757.scala
+++ b/tests/pos/i1757.scala
@@ -1,0 +1,6 @@
+case class B[T](b: List[Int]) {
+  var s: B[Int] = ???
+  def cpy[X](b: List[Int] = b): B[X] = new B[X](b)
+  s.cpy()
+  s.copy()
+}

